src/ch/ch_monitor.c

```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+                virCHProcessStop(CH_DOMAIN_PRIVATE(vm)->driver,
+                                 vm, VIR_DOMAIN_SHUTOFF_SHUTDOWN);
+                virCHDomainObjEndJob(vm);
+            }
+            virObjectUnlock(vm);
+            break;
+            }
+        case virCHMonitorVmEventBooting:
+        case virCHMonitorVmEventPausing:
+        case virCHMonitorVmEventPaused:
+        case virCHMonitorVmEventResuming:
+        case virCHMonitorVmEventSnapshotting:
+        case virCHMonitorVmEventSnapshotted:
+        case virCHMonitorVmEventRestoring:
+        case virCHMonitorVmEventResizing:
+        case virCHMonitorVmEventResized:
+        case virCHMonitorVmEventDeleted:
+        case virCHMonitorVmmEventStarting:
+        case virCHMonitorCpuCreateVcpu:
+            break;
+        case virCHMonitorEventMax:
+        default:
+            VIR_WARN("unkown event from monitor!");
+            break;
+    }
+
+    return 0;
+}
+
+/*
+ * Helper function to find a block of valid JSON
+ * from a stream of multiple JSON blocks.
+ */
+static inline char *end_of_json(char *str)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while (!started || blocks > 0) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return str + i;
+}
+
+/*
+ * Caller should have reference on Monitor and Domain
+ */
+static int virCHMonitorProcessEvents(virCHMonitorPtr mon, int events)
+{
+    virJSONValuePtr obj = NULL;
+    char *buf = mon->buffer;
+    int ret = 0;
+    int i = 0;
+
+    for (i = 0; i < events; i++) {
+        char *end = end_of_json(buf);
+        *end = 0;
+
+        if ((obj = virJSONValueFromString(buf))) {
+            if (virCHMonitorProcessEvent(mon, obj) < 0) {
+                VIR_WARN("Failed to process the event!");
+                ret = -1;
+            }
+            virJSONValueFree(obj);
+        } else {
+            VIR_WARN("Invalid JSON from monitor");
+            ret = -1;
+        }
+
+        *end = '{';
+        buf = end;
+    }
+
+    return ret;
+}
+
+static int virCHMonitorReadProcessEvents(virCHMonitorPtr mon)
+{
+    virDomainObjPtr vm = mon->vm;
+    int events = 0;
+    ssize_t sz = 0;
+    bool incomplete;
+
+    memset(mon->buffer, 0, CH_MONITOR_BUFFER_SZ);
+    do {
+        int ret;
+
+        ret = read(mon->monitor_fd, mon->buffer + sz,
+                   CH_MONITOR_BUFFER_SZ);
+        if (ret < 0) {
+            VIR_WARN("Failed to read monitor events!");
+            return -1;
+        } else if (ret == 0) {
+            g_usleep(G_USEC_PER_SEC);
+            continue;
+        }
+
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer, sz, &incomplete);
+        VIR_DEBUG("Monitor event(%ld):\n%s", sz, mon->buffer);
+    } while (virDomainObjIsActive(vm) && (events == 0 || incomplete));
```

Comment:
So if `virCHMonitorValidateEventsJSON` returns `-1` (which is what `events` get set to) because the json in invalid, `incomplete` isn't set before it is checked here. This would cause a partial read of the monitor fd and likely continue to cause issues handling data from the buffer.

Maybe we should attempt to flush the buffer in case of invalid json?

Reply:
Correct, will fix!

Reply:
Fixed